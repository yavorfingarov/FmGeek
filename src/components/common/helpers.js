export function getStationDisplayName(groupName, stationName) {
    if (groupName) {
        return `${groupName} ${stationName}`;
    } else {
        return stationName;
    }
}
