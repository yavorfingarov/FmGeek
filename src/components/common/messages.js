export const playbackErrorMessage = "Could not play this station.";

export const saveConfirmMessage =
    "This action is irrevisible and might lead to data loss. Make sure you have backup of your data. Continue?";

export const copyErrorMessage = "Could not copy to clipboard.";

export const discardChangesConfirmMessage = "All changes will be lost. Continue?";

export const resetConfirmMessage =
    "You are about to irreversibly reset the application into its initial state. Are you sure?";

export function getParseStationsErrorMessage(groupName, index, property) {
    let message = "Unexpected value";
    if (groupName) {
        message += ` in group '${groupName}'`;
    }
    if (typeof index === "number") {
        message += ` at index ${index}`;
    }
    if (property) {
        message += ` for '${property}'`;
    }
    return message;
}

export function getDuplicateStationNameErrorMessage(stationName) {
    return `Duplicate station name '${stationName}'`;
}

export function getUnsupportedStreamErrorMessage(groupName, stationName) {
    let message = "Unsupported stream";
    if (groupName) {
        message += ` in group '${groupName}'`;
    }
    message += ` for station '${stationName}'`;
    return message;
}
