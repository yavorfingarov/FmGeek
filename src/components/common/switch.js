export class Switch {
    constructor(verb, initialValue) {
        if (initialValue) {
            this[initialValue] = true;
        }
        this[verb] = function (target) {
            for (const key in this) {
                if (key !== verb) {
                    this[key] = false;
                }
            }
            this[target] = true;
        };
    }
}
