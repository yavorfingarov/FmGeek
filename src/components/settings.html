<section x-show="$store.navigation.settings" x-data="settings">
    <h1>Settings</h1>
    <form>
        <fieldset>
            <legend>Sleep Timer</legend>
            <label for="timer">Hours</label>
            <input id="timer" type="time" x-show="!timer" x-model="timeout" />
            <input id="timer" type="time" disabled x-show="timer" :value="timeLeft" />
            <br />
            <button x-text="(timer) ? 'Stop' : 'Start'" @click.prevent="toggleTimer()"></button>
        </fieldset>
        <fieldset>
            <legend>User Interface</legend>
            <label for="theme">Theme</label>
            <select id="theme" x-model="$store.settings.ui.theme">
                <option value="auto">Auto</option>
                <option value="light">Light</option>
                <option value="dark">Dark</option>
            </select>
            <label for="recent-count">Recent stations</label>
            <input
                id="recent-count"
                type="number"
                min="0"
                :max="$store.settings.historyLength"
                x-model.number="$store.settings.ui.recentCount"
            />
        </fieldset>
        <fieldset>
            <legend>Edit</legend>
            <label for="edit">Stations</label>
            <textarea id="edit" class="smaller" spellcheck="false" x-model="json"></textarea>
            <details>
                <summary>Help</summary>
                <p>
                    Depending on your preferences you can have a <code>Station</code> list or a
                    <code>StationGroup</code> list.
                </p>
                <dl>
                    <dt>Station</dt>
                    <dd>
                        <dl>
                            <dt>name</dt>
                            <dd>String</dd>
                            <dt>stream</dt>
                            <dd>String</dd>
                            <dt>website</dt>
                            <dd>String <em>(optional)</em></dd>
                        </dl>
                    </dd>
                    <dt>StationGroup</dt>
                    <dd>
                        <dl>
                            <dt>groupName</dt>
                            <dd>String</dd>
                            <dt>prependToStationName</dt>
                            <dd>Boolean</dd>
                            <dt>stations</dt>
                            <dd>Station[]</dd>
                        </dl>
                    </dd>
                </dl>
            </details>
            <button @click.prevent="save()">Save</button>
            <button @click.prevent="copy()">Copy</button>
            <button @click.prevent="discardChanges()">Discard changes</button>
            <p class="error" x-show="error" x-text="error"></p>
        </fieldset>
        <fieldset>
            <legend>Danger Zone</legend>
            <button @click.prevent="reset()">Factory reset</button>
        </fieldset>
    </form>
</section>
