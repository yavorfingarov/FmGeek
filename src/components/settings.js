import Alpine from "alpinejs";
import {defaultUiSettings, defaultStations, historyLength, defaultTimeout} from "./settings.default";
import {parseStations} from "./settings.import";
import {
    saveConfirmMessage,
    copyErrorMessage,
    discardChangesConfirmMessage,
    resetConfirmMessage
} from "./common/messages";

export function settings() {
    Alpine.store("settings", {
        stations: Alpine.$persist(defaultStations),
        historyLength,
        ui: Alpine.$persist(defaultUiSettings)
    });

    Alpine.data("settings", function () {
        return {
            init() {
                this.json = JSON.stringify(this.$store.settings.stations, null, 2);
                this.jsonDirty = false;
                this.error = null;
            },
            timeout: this.$persist(defaultTimeout),
            timer: null,
            minutesLeft: null,
            checkInput(element) {
                const value = element._x_model.get();
                if (element.min && (!value || value < element.min)) {
                    element._x_model.set(Number(element.min));
                } else if (element.max && value > element.max) {
                    element._x_model.set(Number(element.max));
                } else if (value % 1 !== 0) {
                    element._x_model.set(Math.floor(value));
                }
            },
            toggleTimer() {
                if (this.timer) {
                    this.stopTimer();
                } else if (this.timeout > 0) {
                    this.minutesLeft = this.timeout;
                    this.timer = setInterval(() => this.tick(), 60 * 1000);
                }
            },
            tick() {
                this.minutesLeft--;
                if (this.minutesLeft === 0) {
                    this.stopTimer();
                    this.$dispatch("timer-stop");
                }
            },
            stopTimer() {
                clearInterval(this.timer);
                this.timer = null;
            },
            save() {
                let stations;
                try {
                    stations = parseStations(this.json);
                    this.error = null;
                } catch (error) {
                    this.error = error.message;
                    return;
                }
                if (confirm(saveConfirmMessage)) {
                    this.$store.settings.stations = stations;
                    this.init();
                }
            },
            copy() {
                navigator.clipboard.writeText(this.json).catch(() => {
                    this.error = copyErrorMessage;
                });
            },
            discardChanges() {
                if (confirm(discardChangesConfirmMessage)) {
                    this.init();
                }
            },
            reset() {
                if (confirm(resetConfirmMessage)) {
                    localStorage.clear();
                    history.scrollRestoration = "manual";
                    location.reload();
                }
            }
        };
    });
}
