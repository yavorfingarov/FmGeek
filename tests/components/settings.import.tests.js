import {describe, expect, test} from "vitest";
import {parseStations} from "../../src/components/settings.import";
import {generateStations, generateGroups} from "../common/stations";

describe("parseStations", function () {
    const propertyTestCases = [
        {stationIndex: 0, groupIndex: 2, propertyName: "name", value: null},
        {stationIndex: 3, groupIndex: 2, propertyName: "name", value: ""},
        {stationIndex: 3, groupIndex: 2, propertyName: "name", value: "   "},
        {stationIndex: 3, groupIndex: 2, propertyName: "name", value: 42},
        {stationIndex: 3, groupIndex: 2, propertyName: "name", value: true},
        {stationIndex: 3, groupIndex: 0, propertyName: "name", value: {}},
        {stationIndex: 3, groupIndex: 2, propertyName: "name", value: []},
        {stationIndex: 0, groupIndex: 2, propertyName: "stream", value: null},
        {stationIndex: 3, groupIndex: 2, propertyName: "stream", value: ""},
        {stationIndex: 3, groupIndex: 2, propertyName: "stream", value: "   "},
        {stationIndex: 3, groupIndex: 2, propertyName: "stream", value: 42},
        {stationIndex: 3, groupIndex: 2, propertyName: "stream", value: true},
        {stationIndex: 3, groupIndex: 0, propertyName: "stream", value: {}},
        {stationIndex: 3, groupIndex: 2, propertyName: "stream", value: []},
        {stationIndex: 0, groupIndex: 2, propertyName: "website", value: ""},
        {stationIndex: 3, groupIndex: 2, propertyName: "website", value: "   "},
        {stationIndex: 3, groupIndex: 2, propertyName: "website", value: 42},
        {stationIndex: 3, groupIndex: 2, propertyName: "website", value: true},
        {stationIndex: 3, groupIndex: 0, propertyName: "website", value: {}},
        {stationIndex: 3, groupIndex: 2, propertyName: "website", value: []}
    ];

    const stationTestCases = [
        {stationIndex: 0, groupIndex: 2, value: null},
        {stationIndex: 3, groupIndex: 2, value: ""},
        {stationIndex: 3, groupIndex: 2, value: "   "},
        {stationIndex: 3, groupIndex: 0, value: 42},
        {stationIndex: 3, groupIndex: 2, value: true},
        {stationIndex: 3, groupIndex: 2, value: []}
    ];

    test.each(["", "   ", "{", "undefined", "test"])("throws on invalid JSON (%s)", function (input) {
        expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
    });

    test.each(["null", "{}", "[]"])("throws on invalid root data structure (%s)", function (input) {
        expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
    });

    describe("with grouping", function () {
        test.each(stationTestCases)("throws on invalid groupName ($value)", function (testCase) {
            const groups = generateGroups(5, function (x) {
                x[testCase.groupIndex].groupName = testCase.value;
            });
            const input = JSON.stringify(groups);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test.each(stationTestCases.filter((x) => x.value !== true))(
            "throws on invalid prependToStationName ($value)",
            function (testCase) {
                const groups = generateGroups(5, function (x) {
                    x[testCase.groupIndex].prependToStationName = testCase.value;
                });
                const input = JSON.stringify(groups);
                expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
            }
        );

        test.each(stationTestCases)("throws on invalid stations ($value)", function (testCase) {
            const groups = generateGroups(5, function (x) {
                x[testCase.groupIndex].stations = testCase.value;
            });
            const input = JSON.stringify(groups);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test.each(stationTestCases)("throws on invalid station ($value)", function (testCase) {
            const groups = generateGroups(5, function (x) {
                x[testCase.groupIndex].stations[testCase.stationIndex] = testCase.value;
            });
            const input = JSON.stringify(groups);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test.each(propertyTestCases)("throws on invalid property value ($propertyName = $value)", function (testCase) {
            const groups = generateGroups(5, function (x) {
                x[testCase.groupIndex].stations[testCase.stationIndex][testCase.propertyName] = testCase.value;
            });
            const input = JSON.stringify(groups);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test("throws on unsupported stream", function () {
            const groups = generateGroups(5, function (x) {
                x[2].stations[3].stream = "http://testStream.com";
            });
            const input = JSON.stringify(groups);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test("throws on duplicate station name", function () {
            const stations = [...generateGroups(5), ...generateGroups(1)];
            const input = JSON.stringify(stations);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test("handles valid data", function () {
            const groups = generateGroups(5);
            const input = JSON.stringify(groups);
            const result = parseStations(input);
            expect(result).toStrictEqual(groups);
        });
    });

    describe("without grouping", function () {
        test.each(stationTestCases)("throws on invalid station ($value)", function (testCase) {
            const stations = generateStations(5, function (x) {
                x[testCase.stationIndex] = testCase.value;
            });
            const input = JSON.stringify(stations);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test.each(propertyTestCases)("throws on invalid property value ($propertyName = $value)", function (testCase) {
            const stations = generateStations(5, function (x) {
                x[testCase.stationIndex][testCase.propertyName] = testCase.value;
            });
            const input = JSON.stringify(stations);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test("throws on unsupported stream", function () {
            const stations = generateStations(5, function (x) {
                x[2].stream = "http://testStream.com";
            });
            const input = JSON.stringify(stations);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test("throws on duplicate station name", function () {
            const stations = [...generateStations(5), ...generateStations(1)];
            const input = JSON.stringify(stations);
            expect(() => parseStations(input)).toThrowErrorMatchingSnapshot();
        });

        test("handles valid data", function () {
            const stations = generateStations(5);
            const input = JSON.stringify(stations);
            const result = parseStations(input);
            expect(result).toStrictEqual(stations);
        });
    });
});
