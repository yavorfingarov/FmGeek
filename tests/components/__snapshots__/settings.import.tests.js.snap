// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseStations > throws on invalid JSON (   ) 1`] = `[SyntaxError: Unexpected end of JSON input]`;

exports[`parseStations > throws on invalid JSON () 1`] = `[SyntaxError: Unexpected end of JSON input]`;

exports[`parseStations > throws on invalid JSON ({) 1`] = `[SyntaxError: Expected property name or '}' in JSON at position 1]`;

exports[`parseStations > throws on invalid JSON (test) 1`] = `[SyntaxError: Unexpected token 'e', "test" is not valid JSON]`;

exports[`parseStations > throws on invalid JSON (undefined) 1`] = `[SyntaxError: "undefined" is not valid JSON]`;

exports[`parseStations > throws on invalid root data structure ([]) 1`] = `[Error: Unexpected value]`;

exports[`parseStations > throws on invalid root data structure ({}) 1`] = `[Error: Unexpected value]`;

exports[`parseStations > throws on invalid root data structure (null) 1`] = `[Error: Unexpected value]`;

exports[`parseStations > with grouping > throws on duplicate station name 1`] = `[Error: Duplicate station name 'testGroup0 testName0']`;

exports[`parseStations > with grouping > throws on invalid groupName ('   ') 1`] = `[Error: Unexpected value at index 2 for 'groupName']`;

exports[`parseStations > with grouping > throws on invalid groupName ('') 1`] = `[Error: Unexpected value at index 2 for 'groupName']`;

exports[`parseStations > with grouping > throws on invalid groupName ([]) 1`] = `[Error: Unexpected value at index 2 for 'groupName']`;

exports[`parseStations > with grouping > throws on invalid groupName (42) 1`] = `[Error: Unexpected value at index 0 for 'groupName']`;

exports[`parseStations > with grouping > throws on invalid groupName (null) 1`] = `[Error: Unexpected value at index 2 for 'groupName']`;

exports[`parseStations > with grouping > throws on invalid groupName (true) 1`] = `[Error: Unexpected value at index 2 for 'groupName']`;

exports[`parseStations > with grouping > throws on invalid prependToStationName ('   ') 1`] = `[Error: Unexpected value in group 'testGroup2' for 'prependToStationName']`;

exports[`parseStations > with grouping > throws on invalid prependToStationName ('') 1`] = `[Error: Unexpected value in group 'testGroup2' for 'prependToStationName']`;

exports[`parseStations > with grouping > throws on invalid prependToStationName ([]) 1`] = `[Error: Unexpected value in group 'testGroup2' for 'prependToStationName']`;

exports[`parseStations > with grouping > throws on invalid prependToStationName (42) 1`] = `[Error: Unexpected value in group 'testGroup0' for 'prependToStationName']`;

exports[`parseStations > with grouping > throws on invalid prependToStationName (null) 1`] = `[Error: Unexpected value in group 'testGroup2' for 'prependToStationName']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = '   ') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = '') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = []) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = {}) 1`] = `[Error: Unexpected value in group 'testGroup0' at index 3 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = 42) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = null) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 0 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('name' = true) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'name']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = '   ') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = '') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = []) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = {}) 1`] = `[Error: Unexpected value in group 'testGroup0' at index 3 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = 42) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = null) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 0 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('stream' = true) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'stream']`;

exports[`parseStations > with grouping > throws on invalid property value ('website' = '   ') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'website']`;

exports[`parseStations > with grouping > throws on invalid property value ('website' = '') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 0 for 'website']`;

exports[`parseStations > with grouping > throws on invalid property value ('website' = []) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'website']`;

exports[`parseStations > with grouping > throws on invalid property value ('website' = {}) 1`] = `[Error: Unexpected value in group 'testGroup0' at index 3 for 'website']`;

exports[`parseStations > with grouping > throws on invalid property value ('website' = 42) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'website']`;

exports[`parseStations > with grouping > throws on invalid property value ('website' = true) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3 for 'website']`;

exports[`parseStations > with grouping > throws on invalid station ('   ') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3]`;

exports[`parseStations > with grouping > throws on invalid station ('') 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3]`;

exports[`parseStations > with grouping > throws on invalid station ([]) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3]`;

exports[`parseStations > with grouping > throws on invalid station (42) 1`] = `[Error: Unexpected value in group 'testGroup0' at index 3]`;

exports[`parseStations > with grouping > throws on invalid station (null) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 0]`;

exports[`parseStations > with grouping > throws on invalid station (true) 1`] = `[Error: Unexpected value in group 'testGroup2' at index 3]`;

exports[`parseStations > with grouping > throws on invalid stations ('   ') 1`] = `[Error: Unexpected value in group 'testGroup2' for 'stations']`;

exports[`parseStations > with grouping > throws on invalid stations ('') 1`] = `[Error: Unexpected value in group 'testGroup2' for 'stations']`;

exports[`parseStations > with grouping > throws on invalid stations ([]) 1`] = `[Error: Unexpected value in group 'testGroup2' for 'stations']`;

exports[`parseStations > with grouping > throws on invalid stations (42) 1`] = `[Error: Unexpected value in group 'testGroup0' for 'stations']`;

exports[`parseStations > with grouping > throws on invalid stations (null) 1`] = `[Error: Unexpected value in group 'testGroup2' for 'stations']`;

exports[`parseStations > with grouping > throws on invalid stations (true) 1`] = `[Error: Unexpected value in group 'testGroup2' for 'stations']`;

exports[`parseStations > without grouping > throws on duplicate station name 1`] = `[Error: Duplicate station name 'testName0']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = '   ') 1`] = `[Error: Unexpected value at index 3 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = '') 1`] = `[Error: Unexpected value at index 3 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = []) 1`] = `[Error: Unexpected value at index 3 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = {}) 1`] = `[Error: Unexpected value at index 3 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = 42) 1`] = `[Error: Unexpected value at index 3 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = null) 1`] = `[Error: Unexpected value at index 0 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('name' = true) 1`] = `[Error: Unexpected value at index 3 for 'name']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = '   ') 1`] = `[Error: Unexpected value at index 3 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = '') 1`] = `[Error: Unexpected value at index 3 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = []) 1`] = `[Error: Unexpected value at index 3 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = {}) 1`] = `[Error: Unexpected value at index 3 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = 42) 1`] = `[Error: Unexpected value at index 3 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = null) 1`] = `[Error: Unexpected value at index 0 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('stream' = true) 1`] = `[Error: Unexpected value at index 3 for 'stream']`;

exports[`parseStations > without grouping > throws on invalid property value ('website' = '   ') 1`] = `[Error: Unexpected value at index 3 for 'website']`;

exports[`parseStations > without grouping > throws on invalid property value ('website' = '') 1`] = `[Error: Unexpected value at index 0 for 'website']`;

exports[`parseStations > without grouping > throws on invalid property value ('website' = []) 1`] = `[Error: Unexpected value at index 3 for 'website']`;

exports[`parseStations > without grouping > throws on invalid property value ('website' = {}) 1`] = `[Error: Unexpected value at index 3 for 'website']`;

exports[`parseStations > without grouping > throws on invalid property value ('website' = 42) 1`] = `[Error: Unexpected value at index 3 for 'website']`;

exports[`parseStations > without grouping > throws on invalid property value ('website' = true) 1`] = `[Error: Unexpected value at index 3 for 'website']`;

exports[`parseStations > without grouping > throws on invalid station ('   ') 1`] = `[Error: Unexpected value at index 3]`;

exports[`parseStations > without grouping > throws on invalid station ('') 1`] = `[Error: Unexpected value at index 3]`;

exports[`parseStations > without grouping > throws on invalid station ([]) 1`] = `[Error: Unexpected value at index 3]`;

exports[`parseStations > without grouping > throws on invalid station (42) 1`] = `[Error: Unexpected value at index 3]`;

exports[`parseStations > without grouping > throws on invalid station (null) 1`] = `[Error: Unexpected value at index 0]`;

exports[`parseStations > without grouping > throws on invalid station (true) 1`] = `[Error: Unexpected value at index 3]`;
